src/lib/pubkey/ec_group/ec_inner_data.cpp

```diff
@@ -75,63 +81,122 @@ void EC_Group_Data::set_oid(const OID& oid) {
 }
 
 std::unique_ptr<EC_Scalar_Data> EC_Group_Data::scalar_from_bytes_with_trunc(std::span<const uint8_t> bytes) const {
-   auto bn = BigInt::from_bytes_with_max_bits(bytes.data(), bytes.size(), m_order_bits);
-   return std::make_unique<EC_Scalar_Data_BN>(shared_from_this(), mod_order(bn));
+   if(m_pcurve) {
+      return std::make_unique<EC_Scalar_Data_PC>(shared_from_this(), m_pcurve->scalar_from_bits_with_trunc(bytes));
+   } else {
+      auto bn = BigInt::from_bytes_with_max_bits(bytes.data(), bytes.size(), m_order_bits);
+      return std::make_unique<EC_Scalar_Data_BN>(shared_from_this(), mod_order(bn));
+   }
 }
 
 std::unique_ptr<EC_Scalar_Data> EC_Group_Data::scalar_from_bytes_mod_order(std::span<const uint8_t> bytes) const {
-   BOTAN_ARG_CHECK(bytes.size() <= 2 * order_bytes(), "Input too large");
-   return std::make_unique<EC_Scalar_Data_BN>(shared_from_this(), mod_order(BigInt(bytes)));
+   if(bytes.size() >= 2 * order_bytes()) {
+      return {};
+   }
+
+   if(m_pcurve) {
+      if(auto s = m_pcurve->scalar_from_wide_bytes(bytes)) {
+         return std::make_unique<EC_Scalar_Data_PC>(shared_from_this(), std::move(*s));
+      } else {
+         return {};
+      }
+   } else {
+      return std::make_unique<EC_Scalar_Data_BN>(shared_from_this(), mod_order(BigInt(bytes)));
+   }
 }
 
 std::unique_ptr<EC_Scalar_Data> EC_Group_Data::scalar_random(RandomNumberGenerator& rng) const {
-   return std::make_unique<EC_Scalar_Data_BN>(shared_from_this(), BigInt::random_integer(rng, BigInt::one(), m_order));
+   if(m_pcurve) {
+      return std::make_unique<EC_Scalar_Data_PC>(shared_from_this(), m_pcurve->random_scalar(rng));
+   } else {
+      return std::make_unique<EC_Scalar_Data_BN>(shared_from_this(),
+                                                 BigInt::random_integer(rng, BigInt::one(), m_order));
+   }
 }
 
 std::unique_ptr<EC_Scalar_Data> EC_Group_Data::scalar_zero() const {
-   return std::make_unique<EC_Scalar_Data_BN>(shared_from_this(), BigInt::zero());
+   if(m_pcurve) {
+      return std::make_unique<EC_Scalar_Data_PC>(shared_from_this(), m_pcurve->scalar_zero());
+   } else {
+      return std::make_unique<EC_Scalar_Data_BN>(shared_from_this(), BigInt::zero());
+   }
 }
 
 std::unique_ptr<EC_Scalar_Data> EC_Group_Data::scalar_one() const {
-   return std::make_unique<EC_Scalar_Data_BN>(shared_from_this(), BigInt::one());
+   if(m_pcurve) {
+      return std::make_unique<EC_Scalar_Data_PC>(shared_from_this(), m_pcurve->scalar_one());
+   } else {
+      return std::make_unique<EC_Scalar_Data_BN>(shared_from_this(), BigInt::one());
+   }
 }
 
 std::unique_ptr<EC_Scalar_Data> EC_Group_Data::scalar_from_bigint(const BigInt& bn) const {
-   // Assumed to have been already checked as in range
-   return std::make_unique<EC_Scalar_Data_BN>(shared_from_this(), bn);
+   if(bn <= 0 || bn >= m_order) {
+      return {};
+   }
```

Comment:
The dcoumentation of this method claims:

> This returns a value only if bn is in [0,n) where n is the group order. Otherwise it returns nullptr.

I.e. 0 would be an acceptable input.

Reply:
Fixed